client/powerline.c

```diff
@@ -90,8 +91,18 @@ int main(int argc, char *argv[]) {
 	memset(&server, 0, sizeof(struct sockaddr_un));
 	server.sun_family = AF_UNIX;
 	strncpy(server.sun_path A, address, strlen(address));
+#ifdef __linux__
+	/* We use a socket in the abstract namespace on Linux, and because SUN_LEN uses strlen, it does not work for these. The struct size also fails, so compute manually. */
+	socklen = sizeof(server.sun_family) + strlen(address) A;
+#elif defined SUN_LEN
```

Comment:
Gentoo is using

```C
# define SUN_LEN(ptr) ((size_t) (((struct sockaddr_un *) 0)->sun_path)        \
                      + strlen ((ptr)->sun_path))
```

I think it is broken as well. Wondering what was whoever designed the first part was thinking of. If I am not mistaking this is something like `offsetof` macro, but only usable for `type []` members of the structure. I do not like messing with NULL pointers like this, this always smells like UB even if it has no UB.